lib/spack/spack/build_environment.py

```diff
@@ -130,6 +142,37 @@ def __call__(self, *args, **kwargs):
         return super(MakeExecutable, self).__call__(*args, **kwargs)
 
 
+class ConfigureExecutable(Executable):
+    """Special callable executable object for configure that uses a context
+    manager to swap between frontend and backend target modules. The context
+    manager should only affect platforms that have different architectures
+    on both login and compute nodes. On platforms that have a single
+    architecture (e.g Darwin, Linux), the configure executable works as
+    usual.
+    """
+
+    def __init__(self, name, pkg):
+        super(ConfigureExecutable, self).__init__(name)
+        self.pkg = pkg
+
+    def __call__(self, *args, **kwargs):
+
+        front_end = spack.architecture.front_end_sys_type()
+        platform = self.pkg.architecture.platform
+        frontend_target = platform.target("frontend").module_name
```

Comment:
This is currently specific to modules, but shouldn't it just swap compilers generically?  i.e., it should work on platforms that have a front-end and a back-end, but where the compilers aren't loaded with modules.  Currently this seems like it'll only work on Cray.

Reply:
On the same note, the Cray fix (running `configure` with the frontend environment, then building with the backend compilers) isn't what we'd do for all cross-compiled architectures. e.g., on the Blue Gene machines, you *can* compile a backend executable, but when you run it on the front-end, it actually gets run via RPC on a special partition of backend nodes.  So you *don't* want to do the swap there.  I think whether or not we do the swap should be architecture specific.

Currently the platform classes have a `setup_platform_environment()` method that gets run by `build_environment`.  Shouldn't a change like this be done there?  e.g., shouldn't we pass the module (`m`) to `setup_platform_environment` and have *just* the Cray platform set up this special trick?